core/commands/dag/dag.go

```diff
@@ -241,3 +279,435 @@ var DagResolveCmd = &cmds.Command{
 	},
 	Type: ResolveOutput{},
 }
+
+// It is not guaranteed that a root in a header is actually present in the same ( or any )
+// .car file. This is the case in version 1, and ideally in further versions too
+// Accumulate any root CID seen in a header, and supplement its actual node if/when encountered
+// We will attempt a pin *only* at the end in case all car files were well formed
```

Comment:
> attempt a pin *only* at the end in case all car files were well formed

I can't quite see this in the code. It looks like it's done on a per-root basis, not per-car.

Reply:
Correct - each individual root is opportunistically pinned, to maximize the benefit of what we just ingested being pinned.

Reply:
better change the comment then, it seems misleading talking about the "car files" being "well formed" since you may find one root and pin that and not find another and not pin that.